core-services/workflow/src/main/java/org/silverpeas/core/workflow/engine/WorkflowTools.java

```diff
@@ -0,0 +1,334 @@
+package org.silverpeas.core.workflow.engine;
+
+import org.silverpeas.core.contribution.content.form.Field;
+import org.silverpeas.core.util.logging.SilverLogger;
+import org.silverpeas.core.workflow.api.TaskManager;
+import org.silverpeas.core.workflow.api.UserManager;
+import org.silverpeas.core.workflow.api.WorkflowException;
+import org.silverpeas.core.workflow.api.event.GenericEvent;
+import org.silverpeas.core.workflow.api.instance.Actor;
+import org.silverpeas.core.workflow.api.instance.Question;
+import org.silverpeas.core.workflow.api.instance.UpdatableHistoryStep;
+import org.silverpeas.core.workflow.api.instance.UpdatableProcessInstance;
+import org.silverpeas.core.workflow.api.model.Action;
+import org.silverpeas.core.workflow.api.model.Consequence;
+import org.silverpeas.core.workflow.api.model.Consequences;
+import org.silverpeas.core.workflow.api.model.ProcessModel;
+import org.silverpeas.core.workflow.api.model.QualifiedUsers;
+import org.silverpeas.core.workflow.api.model.State;
+import org.silverpeas.core.workflow.api.model.Trigger;
+import org.silverpeas.core.workflow.api.task.Task;
+import org.silverpeas.core.workflow.api.user.User;
+import org.silverpeas.core.workflow.external.ExternalAction;
+
+import java.util.Hashtable;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A TimeoutRequest indicates the workflow engine that an instance is in an active state since a
+ * too long period
+ */
+class WorkflowTools {
+
+  private WorkflowTools() {
+  }
+
+  /**
+   * Compute the new active states and updates affections
+   * @param event the task done event
+   */
+  public static boolean processAction(UpdatableProcessInstance instance, GenericEvent event,
+      UpdatableHistoryStep step, boolean unactivateResolvedState) throws WorkflowException {
+    // Get the process model
+    ProcessModel model = instance.getProcessModel();
+
+    // Get the task manager
+    TaskManager taskManager = WorkflowHub.getTaskManager();
+
+    // Get the user manager
+    UserManager userManager = WorkflowHub.getUserManager();
+
+    // All active states must be eligible
+    String[] states = instance.getActiveStates();
+    String resolvedStateName = (event.getResolvedState() == null) ? "" : event.getResolvedState().
+        getName();
+    boolean backStatus =
+        (event.getResolvedState() != null) && instance.isStateInBackStatus(resolvedStateName);
+    Hashtable<String, String> oldActiveStates = new Hashtable<>();
+    Hashtable<String, String> eligibleStates = new Hashtable<>();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Replace the synchronized class "Hashtable" by an unsynchronized one such as "HashMap". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1149)